File: src/backend/api/routers/user_feedback.py
"""
User Feedback router - manages general user feedback including feature requests and bug reports.
"""

from typing import List, Optional
from uuid import UUID

from api.auth import User, get_current_user
from api.log import LOGGER
from api.utils import get_user_supabase_client
from fastapi import APIRouter, Depends, HTTPException, Query, Request
from fastapi.security import HTTPAuthorizationCredentials, HTTPBearer
from pydantic import BaseModel, Field
from slowapi import Limiter
from slowapi.util import get_remote_address

# Rate limiting
limiter = Limiter(key_func=get_remote_address)
security_bearer = HTTPBearer()

router = APIRouter(
    prefix="/user-feedback",
    tags=["user-feedback"],
    dependencies=[],
)


# Pydantic models for user_feedback table
class UserFeedbackBase(BaseModel):
    type: str = Field(
        ...,
        description="Type of feedback: feature_request, bug_report, general_feedback",
    )
    text: str = Field(
        ..., description="Feedback content", min_length=10, max_length=2000
    )
    metadata: Optional[dict] = Field(
        default_factory=dict, description="Additional metadata"
    )


class UserFeedbackCreate(UserFeedbackBase):
    pass


class UserFeedbackUpdate(BaseModel):
    text: Optional[str] = Field(None, min_length=10, max_length=2000)
    metadata: Optional[dict] = None


class UserFeedbackResponse(UserFeedbackBase):
    id: UUID
    user_id: UUID
    status: str
    created_at: str
    updated_at: str

    class Config:
        from_attributes = True


# User Feedback endpoints
@router.get("/", response_model=List[UserFeedbackResponse])
@limiter.limit("30/minute")
async def get_user_feedback(
    request: Request,
    current_user: User = Depends(get_current_user),
    type_filter: Optional[str] = Query(None, description="Filter by feedback type"),
    status_filter: Optional[str] = Query(None, description="Filter by status"),
    limit: int = Query(50, le=100, description="Limit number of results"),
    offset: int = Query(0, ge=0, description="Offset for pagination"),
    credentials: HTTPAuthorizationCredentials = Depends(security_bearer),
):
    """Get user feedback for the current user with optional filtering."""
    try:
        user_supabase = get_user_supabase_client(credentials.credentials)
        query = (
            user_supabase.table("user_feedback")
            .select("*")
            .eq("user_id", current_user.id)
        )

        if type_filter:
            query = query.eq("type", type_filter)
        if status_filter:
            query = query.eq("status", status_filter)

        result = (
            query.order("created_at", desc=True)
            .range(offset, offset + limit - 1)
            .execute()
        )

        return [UserFeedbackResponse(**item) for item in result.data]
    except Exception as e:
        LOGGER.error(f"Error fetching user feedback for user {current_user.id}: {e}")
        raise HTTPException(status_code=500, detail="Failed to fetch user feedback")


@router.post("/", response_model=UserFeedbackResponse)
@limiter.limit("5/minute")
async def create_user_feedback(
    request: Request,
    feedback: UserFeedbackCreate,
    current_user: User = Depends(get_current_user),
    credentials: HTTPAuthorizationCredentials = Depends(security_bearer),
):
    """Create new user feedback."""
    try:
        user_supabase = get_user_supabase_client(credentials.credentials)
        # Validate feedback type
        valid_types = [
            "feature_request",
            "bug_report",
            "general_feedback",
            "feeling_feedback",
        ]
        if feedback.type not in valid_types:
            raise HTTPException(
                status_code=400,
                detail=f"Invalid feedback type. Must be one of: {', '.join(valid_types)}",
            )

        data = feedback.model_dump(exclude_unset=True)
        data["user_id"] = current_user.id
        data["status"] = "open"  # Default status

        result = user_supabase.table("user_feedback").insert(data).execute()

        if not result.data:
            raise HTTPException(
                status_code=500, detail="Failed to create user feedback"
            )

        LOGGER.info(
            f"User feedback created: {result.data[0]['id']} by user {current_user.id}"
        )
        return UserFeedbackResponse(**result.data[0])
    except HTTPException:
        raise
    except Exception as e:
        LOGGER.error(f"Error creating user feedback for user {current_user.id}: {e}")
        raise HTTPException(status_code=500, detail="Failed to create user feedback")


@router.get("/{feedback_id}", response_model=UserFeedbackResponse)
@limiter.limit("30/minute")
async def get_user_feedback_by_id(
    request: Request,
    feedback_id: UUID,
    current_user: User = Depends(get_current_user),
    credentials: HTTPAuthorizationCredentials = Depends(security_bearer),
):
    """Get specific user feedback by ID."""
    try:
        user_supabase = get_user_supabase_client(credentials.credentials)
        result = (
            user_supabase.table("user_feedback")
            .select("*")
            .eq("id", feedback_id)
            .eq("user_id", current_user.id)
            .execute()
        )

        if not result.data:
            raise HTTPException(status_code=404, detail="Feedback not found")

        return UserFeedbackResponse(**result.data[0])
    except HTTPException:
        raise
    except Exception as e:
        LOGGER.error(
            f"Error fetching feedback {feedback_id} for user {current_user.id}: {e}"
        )
        raise HTTPException(status_code=500, detail="Failed to fetch feedback")


@router.put("/{feedback_id}", response_model=UserFeedbackResponse)
@limiter.limit("10/minute")
async def update_user_feedback(
    request: Request,
    feedback_id: UUID,
    feedback_update: UserFeedbackUpdate,
    current_user: User = Depends(get_current_user),
    credentials: HTTPAuthorizationCredentials = Depends(security_bearer),
):
    """Update existing user feedback."""
    try:
        user_supabase = get_user_supabase_client(credentials.credentials)
        # Check if feedback exists and belongs to user
        existing = (
            user_supabase.table("user_feedback")
            .select("id")
            .eq("id", feedback_id)
            .eq("user_id", current_user.id)
            .execute()
        )

        if not existing.data:
            raise HTTPException(status_code=404, detail="Feedback not found")

        data = feedback_update.model_dump(exclude_unset=True, exclude_none=True)

        if not data:
            raise HTTPException(status_code=400, detail="No data provided for update")

        result = (
            user_supabase.table("user_feedback")
            .update(data)
            .eq("id", feedback_id)
            .eq("user_id", current_user.id)
            .execute()
        )

        if not result.data:
            raise HTTPException(status_code=500, detail="Failed to update feedback")

        return UserFeedbackResponse(**result.data[0])
    except HTTPException:
        raise
    except Exception as e:
        LOGGER.error(
            f"Error updating feedback {feedback_id} for user {current_user.id}: {e}"
        )
        raise HTTPException(status_code=500, detail="Failed to update feedback")


@router.delete("/{feedback_id}")
@limiter.limit("5/minute")
async def delete_user_feedback(
    request: Request,
    feedback_id: UUID,
    current_user: User = Depends(get_current_user),
    credentials: HTTPAuthorizationCredentials = Depends(security_bearer),
):
    """Delete user feedback."""
    try:
        user_supabase = get_user_supabase_client(credentials.credentials)
        # Check if feedback exists and belongs to user
        existing = (
            user_supabase.table("user_feedback")
            .select("id")
            .eq("id", feedback_id)
            .eq("user_id", current_user.id)
            .execute()
        )

        if not existing.data:
            raise HTTPException(status_code=404, detail="Feedback not found")

        user_supabase.table("user_feedback").delete().eq("id", feedback_id).eq(
            "user_id", current_user.id
        ).execute()

        return {"message": "Feedback deleted successfully"}
    except HTTPException:
        raise
    except Exception as e:
        LOGGER.error(
            f"Error deleting feedback {feedback_id} for user {current_user.id}: {e}"
        )
        raise HTTPException(status_code=500, detail="Failed to delete feedback")

File: src/app/services/userFeedback.ts
import { apiClient } from './api';
import { APP_VERSION } from '@/constants/version';

// General User Feedback types
export interface UserFeedback {
  id: string;
  user_id: string;
  type: 'feature_request' | 'bug_report' | 'general_feedback' | 'feeling_feedback';
  text: string;
  status: 'open' | 'in_progress' | 'resolved' | 'closed';
  metadata?: Record<string, any>;
  created_at: string;
  updated_at: string;
}

export interface CreateUserFeedback {
  type: 'feature_request' | 'bug_report' | 'general_feedback' | 'feeling_feedback';
  text: string;
  metadata?: Record<string, any>;
}

export interface UpdateUserFeedback {
  text?: string;
  metadata?: Record<string, any>;
}

export interface UserFeedbackFilters {
  type?: 'feature_request' | 'bug_report' | 'general_feedback' | 'feeling_feedback';
  status?: 'open' | 'in_progress' | 'resolved' | 'closed';
  limit?: number;
  offset?: number;
}

// User Feedback Service
export class UserFeedbackService {
  // Create new user feedback
  async createFeedback(feedback: CreateUserFeedback): Promise<UserFeedback> {
    try {
      const result = await apiClient.post<UserFeedback>('/v1/user-feedback/', feedback);
      return result;
    } catch (error) {
      console.error('Error creating user feedback:', error);
      throw new Error('Failed to submit feedback. Please try again.');
    }
  }

  // Get user feedback with optional filters
  async getUserFeedback(filters?: UserFeedbackFilters): Promise<UserFeedback[]> {
    try {
      const params = new URLSearchParams();

      if (filters?.type) params.append('type_filter', filters.type);
      if (filters?.status) params.append('status_filter', filters.status);
      if (filters?.limit) params.append('limit', filters.limit.toString());
      if (filters?.offset) params.append('offset', filters.offset.toString());

      const endpoint = params.toString()
        ? `/v1/user-feedback/?${params.toString()}`
        : '/v1/user-feedback/';

      const result = await apiClient.get<UserFeedback[]>(endpoint);
      return result;
    } catch (error) {
      console.error('Error fetching user feedback:', error);
      throw new Error('Failed to load feedback history.');
    }
  }

  // Get specific feedback by ID
  async getFeedbackById(feedbackId: string): Promise<UserFeedback> {
    try {
      const result = await apiClient.get<UserFeedback>(`/v1/user-feedback/${feedbackId}`);
      return result;
    } catch (error) {
      console.error('Error fetching feedback by ID:', error);
      throw new Error('Failed to load feedback.');
    }
  }

  // Update existing feedback
  async updateFeedback(feedbackId: string, updates: UpdateUserFeedback): Promise<UserFeedback> {
    try {
      const result = await apiClient.put<UserFeedback>(`/v1/user-feedback/${feedbackId}`, updates);
      return result;
    } catch (error) {
      console.error('Error updating user feedback:', error);
      throw new Error('Failed to update feedback.');
    }
  }

  // Delete feedback
  async deleteFeedback(feedbackId: string): Promise<void> {
    try {
      await apiClient.delete(`/v1/user-feedback/${feedbackId}`);
    } catch (error) {
      console.error('Error deleting user feedback:', error);
      throw new Error('Failed to delete feedback.');
    }
  }

  // Helper method to create feedback with device/app metadata
  async submitFeedbackWithMetadata(
    type: 'feature_request' | 'bug_report' | 'general_feedback' | 'feeling_feedback',
    text: string,
    additionalMetadata?: Record<string, any>
  ): Promise<UserFeedback> {
    const metadata = {
      app_version: APP_VERSION,
      platform: 'mobile',
      timestamp: new Date().toISOString(),
      ...additionalMetadata,
    };

    return this.createFeedback({
      type,
      text,
      metadata,
    });
  }
}

// Export singleton instance
export const userFeedbackService = new UserFeedbackService();